cmd/kubeadm/app/cmd/join.go

```diff
@@ -378,6 +410,28 @@ func newJoinData(cmd *cobra.Command, args []string, options *joinOptions, out io
 	}, nil
 }
 
+func (j *joinData) Cfg() *kubeadmapi.JoinConfiguration {
+	return j.cfg
+}
+
+func (j *joinData) InitCfg() (*kubeadmapi.InitConfiguration, *clientcmdapi.Config, error) {
+	if j.initCfg != nil && j.tlsBootstrapCfg != nil {
+		return j.initCfg, j.tlsBootstrapCfg, nil
+	}
+	var err error
+	klog.V(1).Infoln("[join] Fetching init configuration")
+	j.initCfg, j.tlsBootstrapCfg, err = fetchInitConfigurationFromJoinConfiguration(j.cfg)
```

Comment:
@neolit123 
I think that the refactoring we are starting for introducing phases is a good opportunity to refactor this function, because `fetchInitConfigurationFromJoinConfiguration` that returns init cfg and tlsBootstrapCfg is kind of weird IMO

I think that we should have two separated methods

- `func (j *joinData) TlsBootstrapCfg() (*clientcmdapi.Config, error)` doing the discovery and sets the private `j.tlsBootstrapCfg` field
- `func (j *joinData) InitCfg() (*kubeadmapi.InitConfiguration, *clientcmdapi.Config, error)`, that uses j.TlsBootstrapCfg() 

Pushing things further, we can also change `fetchInitConfiguration` in fetching only the ClusterConfiguration (and not the ComponentConfigs, so we start addressing the "without kube-proxy" scenario as well

I let you choose what part of this should be part of this PR and what should be instead added to the tracking issue as a separated task

Reply:
> Pushing things further, we can also change fetchInitConfiguration in fetching only the ClusterConfiguration (and not the ComponentConfigs), so we start addressing the "without kube-proxy" scenario as well

> I let you choose what part of this should be part of this PR and what should be instead added to the tracking issue as a separated task

i will leave this for a separate PR.

Reply:
> I think that we should aim to have two separated methods
done.